File: backend/app/models/context.py
from datetime import date
from typing import Literal, Optional

from pydantic import BaseModel, Field


class Workload(BaseModel):
    analysts: int = Field(ge=1)
    developers: int = Field(ge=1)
    testers: int = Field(ge=1)


class ProjectItem(BaseModel):
    name: str
    status: Literal["Завершено", "В работе", "Не начато"]
    start_date: date
    end_date: date
    workload: Workload
    constraints: Optional[str] = None
    deviations: Optional[str] = None
    description: Optional[str] = None


class OrientationItem(BaseModel):
    vision: str
    priority: Literal["Высокий", "Средний", "Низкий"]


class DependencyItem(BaseModel):
    main_project_name: str
    dependent_project_name: str


class PortfolioConstraints(BaseModel):
    analysts_limit: int = Field(ge=0)
    developers_limit: int = Field(ge=0)
    testers_limit: int = Field(ge=0)
    critical_deadline: date


class ContextSubmitRequest(BaseModel):
    orientations: list[OrientationItem] = Field(min_length=1)
    projects: list[ProjectItem] = Field(min_length=1)
    dependencies: list[DependencyItem] = []
    portfolio_constraints: PortfolioConstraints

File: frontend/src/app/api/context.ts
import { api } from './client';

export type ContextPayload = {
  orientations: { vision: string; priority: 'Высокий' | 'Средний' | 'Низкий' }[];
  projects: {
    name: string;
    status: 'Завершено' | 'В работе' | 'Не начато';
    start_date: string;
    end_date: string;
    workload: { analysts: number; developers: number; testers: number };
    constraints?: string;
    deviations?: string;
    description?: string;
  }[];
  dependencies: { main_project_name: string; dependent_project_name: string }[];
  portfolio_constraints: {
    analysts_limit: number;
    developers_limit: number;
    testers_limit: number;
    critical_deadline: string;
  };
};

export type SavedContext = {
  orientations: { vision: string; priority: string }[];
  projects: {
    name: string;
    status: string;
    start_date: string;
    end_date: string;
    workload: { analysts: number; developers: number; testers: number };
    constraints?: string;
    deviations?: string;
    description?: string;
  }[];
  dependencies: { main_project_name: string; dependent_project_name: string }[];
  portfolio_constraints: {
    analysts_limit: number;
    developers_limit: number;
    testers_limit: number;
    critical_deadline: string;
  } | null;
};

export const contextApi = {
  get: (contextId: string) =>
    api.get<SavedContext>(`/contexts/${contextId}/context`),
  submit: (contextId: string, payload: ContextPayload) =>
    api.post<{ detail: string }>(`/contexts/${contextId}/context`, payload),
};
